src/rules/no-incompatible-branch-types.coffee

```diff
@@ -0,0 +1,63 @@
+ts = require 'typescript'
+
+# getSourceFileOfNode = (tsNode) ->
+#   while node and node.kind isnt ts.SyntaxKind.SourceFile
+#     node = node.parent
+#   node
+
+module.exports =
+  meta:
+    docs:
+      description:
+        'Flag incompatible early-return types in branch()/branchPure()'
+      category: 'Possible Errors'
+      recommended: yes
+    schema: []
+
+  create: (context) ->
+    {parserServices} = context
+    return {} unless parserServices?.hasFullTypeInformation
+
+    typeChecker = parserServices.program.getTypeChecker()
+
+    CallExpression: (node) ->
+      {callee, parent} = node
+      return unless (
+        callee?.type is 'Identifier' and
+        callee.name in [### 'renderNothing', ### 'returns']
```

Comment:
`renderNothing()` returns `null` which is compatible with React components, so don't need to worry about it until trying to check (non-React component) `branchPure()`'s